File: cpp/src/test/rktest.cc
// ===== BEGIN FILE =========================================================================
// **                              P R O P R I E T A R Y                                   **
// ==========================================================================================
// **                                                                                      **
// **  Copyright (c) 2006, Stephen W. Soliday                                              **
// **                      stephen@soliday.com                                             **
// **                      http://www.soliday.com/stephen                                  **
// **                                                                                      **
// **  This file, and the associated algorithms, are not free software; you may not        **
// **  redistribute them and/or modify them. These algorithms were developed and           **
// **  implemented for the purpose of an internal assessment and have, as yet, not been    **
// **  publicly distributed. Development of these algorithms have been at the sole cost    **
// **  in both time and funding by their author. Until such a public release is made,      **
// **  the author retains ALL RIGHTS to these algorithms. It is expected that if this      **
// **  program or any of the algorithms contained herein are deemed releasable they will   **
// **  be released under the GNU Public license for non-commercial use and/or with         **
// **  restricted rights for government use. At that time each source file will contain    **
// **  either/both the standard GPL statement/disclaimer, and/or                           **
// **  the DFARS Restricted Rights Legend.                                                 **
// **                                                                                      **
// **  These algorithms exists at the present time WITHOUT ANY WARRANTY; without even      **
// **  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.        **
// **  As you are not supposed to be in possession of this file if you use it,             **
// **  you do so AT YOUR OWN RISK.                                                         **
// **                                                                                      **
// **  ----- Modification History -------------------------------------------------------  **
// **                                                                                      **
// **  Author Stephen W. Soliday                                                           **
// **  Date   2006-11-24                                                                   **
// **                                                                                      **
/**   \file rktest.cc
 *    \brief Test Numerical Integrator.
 *     This program tests the functionality of \a RK4. \a RK4 is a fourth order Runge-Kutta
 *     numerical integrator.
 */
// ==========================================================================================

#include <rk4.hh>
#include <psgraph.hh>
#include <2body.h>

// ==========================================================================================
// ------------------------------------------------------------------------------------------
class Motion :public RK4 {
  // ----------------------------------------------------------------------------------------
public:
  int done;

  Motion( void ) :  RK4(8), done(0) {};
  virtual ~Motion( void ) { };

  virtual void DIFEQ ( double *Qd, double *Q, double t, double *P );
  virtual int  CHECK (             double *Q, double t, double *P );
};

// ==========================================================================================
/** Check State.
 *  Determine if any two bodies are closer than the sum of thier radii.
 *  At a predetermined time, add velocity to the smallest body.
 * \param Q current state vector.
 * \param t current time.
 * \param P parameter vector.
 * \return success=0, failure=non-zero.
 */
// ------------------------------------------------------------------------------------------
int Motion::CHECK( double *Q, double t, double *P ) {
  // ----------------------------------------------------------------------------------------
  double dx = Q[6] - Q[4];
  double dy = Q[7] - Q[5];
  double r2 = P[1] + P[3];

  return (((r2*r2) < ((dx*dx)+(dy*dy))) ? (0) : (1));
}

// ==========================================================================================
/** Integrate.
 *  Integrate for time \a t0 to \a t1. Using a fourth order Runge-Kutta numerical integrator.
 *  The equations of mothion describe a generic two body gravitational problem.
 * \param Qd first time derivative of the current state vector.
 * \param Q current state vector.
 * \param t current time.
 * \param P parameter vector.
 * \return success=0, failure=non-zero
 */
// ------------------------------------------------------------------------------------------
void Motion::DIFEQ( double *Qd, double *Q, double t, double *P ) {
  // ----------------------------------------------------------------------------------------
  if (done == 0) {

    double xd1 = Q[0];  double x1 = Q[4];
    double yd1 = Q[1];  double y1 = Q[5];
    double xd2 = Q[2];  double x2 = Q[6];
    double yd2 = Q[3];  double y2 = Q[7];

    double m1   = P[0];
    double m2   = P[2];

    double dx21 = x2 - x1;
    double dy21 = y2 - y1;
    double dx12 = x1 - x2;
    double dy12 = y1 - y2;

    double r12sq = (dx12*dx12) + (dy12*dy12);
    double den   = POWER(r12sq,1.5);

    if (den > 0.0) {} else { fprintf( stderr, "INFINITY\n" ); }

    double a1 = N_G*m2*dx21;
    double a2 = N_G*m2*dy21;
    double a3 = N_G*m1*dx12;
    double a4 = N_G*m1*dy12;

    Qd[0] = SAFEDIV(a1,den);
    Qd[1] = SAFEDIV(a2,den);
    Qd[2] = SAFEDIV(a3,den);
    Qd[3] = SAFEDIV(a4,den);
    Qd[4] = xd1;
    Qd[5] = yd1;
    Qd[6] = xd2;
    Qd[7] = yd2;
  }
}

// ==========================================================================================
/** Entry Point.
 *  Standard CPP entry point.
 * \param argc number of command line arguments.
 * \param argv list of command line fields.
 * \return success=0, failure=non-zero
 */
// ------------------------------------------------------------------------------------------
int main( int argc, char *argv[] ) {
  // ----------------------------------------------------------------------------------------
  char *progName = argv[0];
  char *fspc     = argv[1];

  if (argc != 2) {
    std::cerr << "USAGE: " << progName << " test.ps" << std::endl << std::endl;
    return 1;
  }

  double param[4];
  double state[8];
  double t  = 0.0;
  double dt = (MAXT * 86400.0) / ((double) ITER);

  param[0] = Me;
  param[1] = Re;
  param[2] = Mm;
  param[3] = Rm;

  state[0] = VXe;       state[4] = Xe;
  state[1] = VYe;       state[5] = Ye;
  state[2] = VXm;       state[6] = Xm;
  state[3] = VYm;       state[7] = Ym;

  double x1 = state[4];
  double y1 = state[5];
  double x2 = state[6];
  double y2 = state[7];

  PSGraph *PSG = new PSGraph( fspc );

  PSG->setWorldCo( -maxr, -maxr, maxr, maxr );
  PSG->initGraphics();

  Motion *MM = new Motion();

  for (int i=0; i<ITER; i++) {

    t = MM->integrate( state, t, t+dt, ISTEP, param );

    PSG->drawLine( x1, y1, state[4], state[5]);
    PSG->drawLine( x2, y2, state[6], state[7] );
 
    x1 = state[4];
    y1 = state[5];
    x2 = state[6];
    y2 = state[7];

    if (MM->done == 1) { break; }
  }

  PSG->drawCircle( state[4], state[5], param[1] );
  PSG->drawCircle( state[6], state[7], param[3] );
  
  delete PSG;

  return 0;
}

// ==========================================================================================
// **                              P R O P R I E T A R Y                                   **
// =========================================================================== END FILE =====

File: fortran/src/test/rktest.f
C ===== BEGIN FILE =============================================================
C **                        P R O P R I E T A R Y                             **
C ==============================================================================
C **                                                                          **
C **  COPYRIGHT (C) 2007, STEPHEN W. SOLIDAY                                  **
C **                      STEPHEN@SOLIDAY.COM                                 **
C **                      HTTP://WWW.SOLIDAY.COM/STEPHEN                      **
C **                                                                          **
C **  THIS FILE, AND THE ASSOCIATED ALGORITHMS, ARE NOT FREE SOFTWARE; YOU    ** 
C **  MAY NOT REDISTRIBUTE THEM AND/OR MODIFY THEM. THESE ALGORITHMS WERE     **
C **  DEVELOPED AND IMPLEMENTED FOR THE PURPOSE OF AN INTERNAL ASSESSMENT AND **
C **  HAVE, AS YET, NOT BEEN PUBLICLY DISTRIBUTED. DEVELOPMENT OF THESE       **
C **  ALGORITHMS HAVE BEEN AT THE SOLE COST IN BOTH TIME AND FUNDING BY THEIR ** 
C **  AUTHOR. UNTIL SUCH A PUBLIC RELEASE IS MADE, THE AUTHOR RETAINS ALL     **
C **  RIGHTS TO THESE ALGORITHMS. IT IS EXPECTED THAT IF THIS PROGRAM OR ANY  **
C **  OF THE ALGORITHMS CONTAINED HEREIN ARE DEEMED RELEASABLE THEY WILL BE   **
C **  RELEASED UNDER THE GNU PUBLIC LICENSE FOR NON-COMMERCIAL USE AND/OR     **
C **  WITH RESTRICTED RIGHTS FOR GOVERNMENT USE. AT THAT TIME EACH SOURCE     **
C **  FILE WILL CONTAIN EITHER/BOTH THE STANDARD GPL STATEMENT/DISCLAIMER,    **
C **  AND/OR THE DFARS RESTRICTED RIGHTS LEGEND.                              **
C **                                                                          **
C **  THESE ALGORITHMS EXISTS AT THE PRESENT TIME WITHOUT ANY WARRANTY;       **
C **  WITHOUT EVEN THE IMPLIED WARRANTY OF MERCHANTABILITY OR FITNESS FOR A   **
C **  PARTICULAR PURPOSE. AS YOU ARE NOT SUPPOSED TO BE IN POSSESSION OF THIS **
C **  FILE IF YOU USE IT, YOU DO SO AT YOUR OWN RISK.                         **
C **                                                                          **
C **  ----- MODIFICATION HISTORY -------------------------------------------- **
C **                                                                          **
C **  AUTHOR STEPHEN W. SOLIDAY                                               **
C **  DATE   2007-02-18                                                       **
C **                                                                          **
C ==============================================================================

C----+==================================================================--------
      BLOCK DATA GLOB01
C----+==================================================================--------
      INTEGER DONE
      COMMON /TPARS/DONE
C     ------------------------------------------------------------------
      DATA DONE/0/
      END

C----+==================================================================--------
      BLOCK DATA TWOBODY
C----+==================================================================--------
      INTEGER ITER,ISTEP
      DOUBLE PRECISION MAXR,MAXT,ME,MM,RE,RM,XE,YE,XM,YM,
     1     VXE,VYE,VXM,VYM
C     ------------------------------------------------------------------
      COMMON /TBPARS/ITER,ISTEP,MAXR,MAXT,ME,MM,RE,RM,XE,YE,XM,YM,
     1     VXE,VYE,VXM,VYM
C     ------------------------------------------------------------------
      DATA ITER/10000/,ISTEP/60/
      DATA MAXR/4.6D+8/
      DATA MAXT/29.0/
      DATA ME/5.9742D+24/
      DATA MM/7.347673D+22/
      DATA RE/6370996.0/
      DATA RM/1737146.0/
      DATA XE/-4.41156D+6/
      DATA YE/0.0/
      DATA XM/3.58692D+8/
      DATA YM/0.0/
      DATA VXE/0.0/
      DATA VYE/-11.7422/
      DATA VXM/0.0/
      DATA VYM/1082.0/
C     ------------------------------------------------------------------
      END

C----+==================================================================--------
C     CHECK STATE.
C     DETERMINE IF ANY TWO BODIES ARE CLOSER THAN THE SUM OF THIER RADII.
C     AT A PREDETERMINED TIME, ADD VELOCITY TO THE SMALLEST BODY.
C    \PARAM Q CURRENT STATE VECTOR.
C    \PARAM T CURRENT TIME.
C    \PARAM P PARAMETER VECTOR.
C    \RETURN SUCCESS=0, FAILURE=NON-ZERO.
C----+==================================================================--------
      FUNCTION CHECK(Q,T,P)
      INTEGER CHECK
C----+==================================================================--------
      DOUBLE PRECISION Q(*),T,P(*)
C     ------------------------------------------------------------------
      DOUBLE PRECISION DX,DY,DR,S
C     ------------------------------------------------------------------
      INTEGER DONE
      COMMON /TPARS/DONE
C     ------------------------------------------------------------------
      DX=Q(7)-Q(5)
      DY=Q(8)-Q(6)
      DR=P(2)+P(4)

      S=(DX*DX)+(DY*DY)-(DR*DR)

      IF(S.LT.0.0) THEN
         DONE=1
         CHECK=1
      ELSE           
         CHECK=0
      END IF
C     ------------------------------------------------------------------
      RETURN
      END

C----+==================================================================--------
C     INTEGRATE FOR TIME \A T0 TO \A T1. USING A FOURTH ORDER RUNGE-KUTTA 
C     NUMERICAL INTEGRATOR. THE EQUATIONS OF MOTION DESCRIBE A GENERIC TWO BODY
C     GRAVITATIONAL PROBLEM.
C    \PARAM QD FIRST TIME DERIVATIVE OF THE CURRENT STATE VECTOR.
C    \PARAM Q CURRENT STATE VECTOR.
C    \PARAM T CURRENT TIME.
C    \PARAM P PARAMETER VECTOR.
C----+==================================================================--------
      SUBROUTINE DIFEQ(QD,Q,T,P)
C----+==================================================================--------
      DOUBLE PRECISION QD(*),Q(*),T,P(*)
C     ------------------------------------------------------------------
      DOUBLE PRECISION N2PI,NG
      COMMON /CONSTANTS/N2PI,NG
C     ------------------------------------------------------------------
      INTEGER DONE
      COMMON /TPARS/DONE
C     ------------------------------------------------------------------
      DOUBLE PRECISION XD1,YD1,XD2,YD2,M1,M2,DX21,DY21,DX12,DY12
      DOUBLE PRECISION R12SQ,DEN,A1,A2,A3,A4,X1,Y1,X2,Y2
C     ------------------------------------------------------------------
      IF(1.EQ.DONE) GOTO 99

      XD1   = Q(1)
      YD1   = Q(2)
      XD2   = Q(3)
      YD2   = Q(4)
      X1    = Q(5)
      Y1    = Q(6)
      X2    = Q(7)
      Y2    = Q(8)

      M1    = P(1)
      M2    = P(3)

      DX21  = X2-X1
      DY21  = Y2-Y1
      DX12  = X1-X2
      DY12  = Y1-Y2

      R12SQ = (DX12*DX12) + (DY12*DY12);

      DEN   = R12SQ**1.5
      
      A1 = NG*M2*DX21;
      A2 = NG*M2*DY21;
      A3 = NG*M1*DX12;
      A4 = NG*M1*DY12;
      

      IF (DEN.LE.0.0) THEN
         WRITE(0,*) 'INFINITY'
      END IF

      QD(1) = SAFEDIV(A1,DEN)
      QD(2) = SAFEDIV(A2,DEN)
      QD(3) = SAFEDIV(A3,DEN)
      QD(4) = SAFEDIV(A4,DEN)

      QD(5) = XD1
      QD(6) = YD1
      QD(7) = XD2
      QD(8) = YD2

 99   CONTINUE
C     ------------------------------------------------------------------
      RETURN
      END

C----+==================================================================--------
C     ENTRY POINT.
C----+==================================================================--------
      PROGRAM RKTEST
C----+==================================================================--------
      INTEGER DONE
      COMMON /TPARS/DONE
C     ------------------------------------------------------------------
      DOUBLE PRECISION N2PI,NG
      COMMON /CONSTANTS/N2PI,NG
C     ------------------------------------------------------------------
      INTEGER ITER,ISTEP
      DOUBLE PRECISION MAXR,MAXT,ME,MM,RE,RM,XE,YE,XM,YM,
     1     VXE,VYE,VXM,VYM
C     ------------------------------------------------------------------
      COMMON /TBPARS/ITER,ISTEP,MAXR,MAXT,ME,MM,RE,RM,XE,YE,XM,YM,
     1     VXE,VYE,VXM,VYM
C     ------------------------------------------------------------------
      DOUBLE PRECISION PARAM, STATE
      DIMENSION PARAM(4), STATE(8)
      INTEGER I
      DOUBLE PRECISION T,DT,X1,X2,Y1,Y2
C     ------------------------------------------------------------------
      T  = 0.0
      DT = (MAXT*86400.0) / DBLE(ITER)

      X1 = XE
      Y1 = YE
      X2 = XM
      Y2 = YM

      PARAM(1) = ME
      PARAM(2) = RE
      PARAM(3) = MM
      PARAM(4) = RM
      
      STATE(1) = VXE
      STATE(2) = VYE
      STATE(3) = VXM
      STATE(4) = VYM
      STATE(5) = XE
      STATE(6) = YE
      STATE(7) = XM
      STATE(8) = YM

      CALL SETWORLDCO(-MAXR,-MAXR,MAXR,MAXR)
      CALL INITGRAPHICS

      DONE = 0

      DO 10 I=1,ITER
         T = RK4(STATE,8,T,T+DT,ISTEP,PARAM)

         CALL DRAWLINE(X1,Y1,STATE(5),STATE(6))
         CALL DRAWLINE(X2,Y2,STATE(7),STATE(8))
         
         X1 = STATE(5)
         Y1 = STATE(6)
         X2 = STATE(7)
         Y2 = STATE(8)

         IF(1.EQ.DONE) GOTO 20
 10   CONTINUE
 20   CONTINUE
      
      CALL DRAWCIRCLE(STATE(5),STATE(6),PARAM(2))
      CALL DRAWCIRCLE(STATE(7),STATE(8),PARAM(4))
  
      CALL PRINTGRAPHICS

C     ------------------------------------------------------------------
      STOP
      END

C ==============================================================================
C **                        P R O P R I E T A R Y                             **
C =============================================================== END FILE =====
